pkg/helm/metallb.go

```diff
@@ -0,0 +1,445 @@
+	if err != nil {
+		return nil, err
+	}
+	chart.chart, err = loader.Load(chartPath)
+	if err != nil {
+		return nil, err
+	}
+	chart.config, err = loadMetalLBConfig(namespace, isOpenshift)
+	if err != nil {
+		return nil, err
+	}
+	return chart, nil
+}
+
+func overrideControllerParameters(crdConfig *metallbv1beta1.MetalLB, obj *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	controllerConfig := crdConfig.Spec.ControllerConfig
+	if controllerConfig == nil || !isControllerDeployment(obj) {
+		return obj, nil
+	}
+	var controller *appsv1.Deployment
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.Object, &controller)
+	if err != nil {
+		return nil, err
+	}
+	if controllerConfig.Affinity != nil {
+		controller.Spec.Template.Spec.Affinity = controllerConfig.Affinity
+	}
+	for j, container := range controller.Spec.Template.Spec.Containers {
+		if container.Name == "controller" && controllerConfig.Resources != nil {
+			controller.Spec.Template.Spec.Containers[j].Resources = *controllerConfig.Resources
+		}
+	}
+	objMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(controller)
+	if err != nil {
+		return nil, err
+	}
+	return &unstructured.Unstructured{Object: objMap}, nil
+}
+
+func overrideSpeakerParameters(crdConfig *metallbv1beta1.MetalLB, obj *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	speakerConfig := crdConfig.Spec.SpeakerConfig
+	if speakerConfig == nil || !isSpeakerDaemonSet(obj) {
+		return obj, nil
+	}
+	var speaker *appsv1.DaemonSet
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.Object, &speaker)
+	if err != nil {
+		return nil, err
+	}
+	if speakerConfig.Affinity != nil {
+		speaker.Spec.Template.Spec.Affinity = speakerConfig.Affinity
+	}
+	for j, container := range speaker.Spec.Template.Spec.Containers {
+		if container.Name == "speaker" && speakerConfig.Resources != nil {
+			speaker.Spec.Template.Spec.Containers[j].Resources = *speakerConfig.Resources
+		}
+	}
+	objMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(speaker)
+	if err != nil {
+		return nil, err
+	}
+	return &unstructured.Unstructured{Object: objMap}, nil
+}
+
+func isControllerDeployment(obj *unstructured.Unstructured) bool {
+	return obj.GetKind() == "Deployment" && obj.GetName() == "controller"
+}
+
+func isSpeakerDaemonSet(obj *unstructured.Unstructured) bool {
+	return obj.GetKind() == "DaemonSet" && obj.GetName() == "speaker"
+}
+
+func isServiceMonitor(obj *unstructured.Unstructured) bool {
+	return obj.GetKind() == "ServiceMonitor"
+}
+
+func (c *mlbChartConfig) patchToChartValues(crdConfig *metallbv1beta1.MetalLB, withPrometheus bool, valueMap map[string]interface{}) {
+	c.withGeneralValues(crdConfig, valueMap)
+	c.withPrometheusValues(valueMap)
+	c.withControllerValues(crdConfig, valueMap)
+	c.withSpeakerValues(crdConfig, valueMap)
+}
+
+func (c *mlbChartConfig) withGeneralValues(crdConfig *metallbv1beta1.MetalLB, valueMap map[string]interface{}) {
```

Comment:
I know it was there before, but with fresh eyes this whole patch / with pattern is not really ideal. How about we switch to a append style like:

valueMap = fillGeneralValues(valueMap, crdConfig) ?

Also, I'd rename to valuesMap

Reply:
Also, this (and probably the others too, with a bit more effort) can be demoted to functions instead of methods

Reply:
changed to something a bit different where the methods return the object and we set the key of the map from outside. also, the With* were functions before and I moved them to methods as I think it makes them easier to read and call (also allows sharing names between the charts like prometheus)